### DISCUSSION TITLE
Help and tips on how to implement an algorithm in jax

### DISCUSSION THREAD

**Discussion by elientumba2019** [jax-ml/jax#21658] [Q&A] [answered]
![image](https://github.com/google/jax/assets/26796538/fb7d2eaa-4b3b-44e4-b37c-dac02355a732)

Hello, 

I am wondering how i might go about implementing efficiently the random walk algorithm above  using jax. 

One way i thought about was to vmap a tensor of positions (N, 3) then in the vmaped function use a lax.while with the given conditions. This however is slow especially for multiple walks. What i would ideally want to do is to walk all my points simultaneously. This can easily be done with lax.for_i however it is not very straighforward to enforce the conditions on each element of the tensor when implementing in this way. 

Any help and tips would be greatly appreciated.
Thank you.

**Comment by jakevdp** [comment, marked as answer]
Hi - thanks for the question! In general JAX is not going to be particularly performant for loopy algorithms like this one. JAX works best with vectorized operations on accelerators, and iterative algorithms explicitly disallow the kinds of implicit parallelism that makes operations on GPU and TPU fast.

That said, if I were implementing this algorithm I'd use [`jax.lax.while_loop`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.while_loop.html) to implement the logic for a single value, and then use `vmap` to efficiently create a batched version. Then you don't have to worry about manually tracking the conditions of each element in the tensor, and the `vmap` solution will be as fast as any hand-written batched version.
